cf-agent/verify_files.c

```diff
@@ -651,12 +646,57 @@ static PromiseResult RenderTemplateMustache(EvalContext *ctx, const Promise *pp,
         cfPS(ctx, LOG_LEVEL_ERR, PROMISE_RESULT_FAIL, pp, a, "Error rendering mustache template '%s'", a.edit_template);
         result = PromiseResultUpdate(result, PROMISE_RESULT_FAIL);
         JsonDestroy(default_template_data);
-        WriterClose(template_writer);
         BufferDestroy(output_buffer);
         return PromiseResultUpdate(result, PROMISE_RESULT_FAIL);
     }
 }
 
+static PromiseResult RenderTemplateMustacheFromFIle(EvalContext *ctx, const Promise *pp, Attributes a,
+                                            EditContext *edcontext)
+{
+    PromiseResult result = PROMISE_RESULT_NOOP;
+
+    if (!FileCanOpen(a.edit_template, "r"))
+    {
+        cfPS(ctx, LOG_LEVEL_ERR, PROMISE_RESULT_FAIL, pp, a, "Template file '%s' could not be opened for reading", a.edit_template);
+        return PromiseResultUpdate(result, PROMISE_RESULT_FAIL);
+    }
+
+
+    int template_fd = safe_open(a.edit_template, O_RDONLY | O_TEXT);
+    Writer *template_writer = NULL;
+    if (template_fd >= 0)
+    {
+        template_writer = FileReadFromFd(template_fd, SIZE_MAX, NULL);
+        close(template_fd);
+    }
+    if (!template_writer)
+    {
+        cfPS(ctx, LOG_LEVEL_ERR, PROMISE_RESULT_FAIL, pp, a, "Could not read template file '%s'", a.edit_template);
+        return PromiseResultUpdate(result, PROMISE_RESULT_FAIL);
+    }
+
+    result =  RenderTemplateMustache(ctx, pp, a, edcontext, (char *)StringWriterData(template_writer));
```

Comment:
No cast for `void *` parameters.

Reply:
Is not a void, but `const char *`
```
verify_files.c:582:22: note: expected 'char *' but argument is of type 'const char *'
```

So I need to cast it else we get a warning

Reply:
This function is used twice and this function requires a char * parameter and the function StringWriiterData returns const char *.  see also RenderTemplateMustacheFromString

Reply:
Yes, change the parameter (in function definition) of `RenderTemplateMustache()`, it does not modify the contents of `template`, so it can be const:
```
static PromiseResult RenderTemplateMustache(EvalContext *ctx, const Promise *pp, Attributes a,
                                            EditContext *edcontext, const char *template)
```
It can then accept both `char *` and `const char *`, so remove the cast:
```
    result =  RenderTemplateMustache(ctx, pp, a, edcontext, StringWriterData(template_writer));
```
(You shouldn't discard the const which `StringWriterData()` returns).

I have tested this on your branch, and it works.